File: 2024/March/3-16 525 Contiguous Array/ContiguousArray.java
import java.util.HashMap;

class Solution {
    public int findMaxLength(int[] nums) {
        int[] sums = new int[nums.length];
        int acc = 0;
        HashMap<Integer, Integer> firstInstance = new HashMap<Integer, Integer>();
        int maxZero = -1;
        int maxInterval = 0;

        for(int i = 0; i < nums.length; i++) {
            if(nums[i] == 0) {
                acc--;
            }
            else {
                acc++;
            }

            if(acc == 0) {
                maxZero = i;
            }
            else {
                if(!firstInstance.containsKey(acc)) {
                    firstInstance.put(acc, i);
                }
                else {
                    if(i - firstInstance.get(acc) > maxInterval) {
                        maxInterval = i - firstInstance.get(acc);
                    }
                }
            }

            sums[i] = acc;
        }

        if(maxZero > maxInterval) {
            return maxZero + 1;
        }

        return maxInterval;
    }
}

File: 2024/March/3-16 525 Contiguous Array/ContiguousArray.py
from typing import List

class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        sums = [0] * len(nums)
        acc = 0
        firstInstance = {}
        maxZero = -1
        maxInterval = 0

        for i in len(nums):
            if nums[i] == 0:
                acc -= 1
            else:
                acc += 1

            if acc == 0:
                maxZero = i
            else:
                if acc not in firstInstance:
                    firstInstance[acc] = i
                else:
                    if i - firstInstance[acc] > maxInterval:
                        maxInterval = i - firstInstance[acc]
            
            sums[i] = acc
        
        if maxZero > maxInterval:
            return maxZero + 1
        
        return maxInterval